sessionctx/variable/session.go

```diff
@@ -130,6 +130,18 @@ func (r *retryInfoAutoIDs) getCurrent() (int64, bool) {
 	return id, true
 }
 
+// SavepointRecord is a (savepoint-name, CP) pair.
+type SavepointRecord struct {
+	// name is the name of the savepoint
+	Name string
+	// Savepoint is the transaction savepoint.
+	Savepoint interface{}
+	// TableDeltaMap is the savepoint of TransactionContext.TableDeltaMap
+	TableDeltaMap map[int64]TableDelta
+	// pessimisticLockCache is the savepoint of TransactionContext.pessimisticLockCache
+	pessimisticLockCache map[string][]byte
```

Comment:
I'm a bit afraid it would be a bit harder to maintain `TxnCtx` in the future, since some of the information needs to be restored when rolling back to a savepoint while some does not. Could it be better to split TxnCtx into two parts: those that should be restored and those should not. So anyone who will add new information to it can easily put them in the right place. 
Then we can mechanically clone the first part here, instead of specifying each field.

While this may be overengineering. What do you think?

Reply:
Nice catch, I'm a little worried about that too.